utils/fs/memory/memory.go

```diff
@@ -0,0 +1,333 @@
+			return path
+		}
+
+		return ""
+	}
+
+	if path[0] == separator {
+		path = path[1:]
+	}
+
+	l := len(path)
+	if path[l-1] == separator {
+		path = path[:l-1]
+	}
+
+	return path
+}
+
+func (fs *Memory) Join(elem ...string) string {
+	return filepath.Join(elem...)
+}
+
+func (fs *Memory) Dir(path string) fs.Filesystem {
+	return &Memory{
+		base: fs.Join(fs.base, path),
+		s:    fs.s,
+	}
+}
+
+func (fs *Memory) Base() string {
+	return fs.base
+}
+
+type file struct {
+	fs.BaseFile
+
+	c     *content
+	p     int64
+	flags int
+}
+
+func newFile(base, fullpath string, flags int) *file {
+	filename, _ := filepath.Rel(base, fullpath)
+
+	return &file{
+		BaseFile: fs.BaseFile{BaseFilename: filename},
+		c:        &content{},
+		flags:    flags,
+	}
+}
+
+func (f *file) Read(b []byte) (int, error) {
+	if f.IsClosed() {
+		return 0, fs.ErrClosed
+	}
+
+	if f.flags&os.O_RDWR != 0 && f.flags&os.O_RDONLY != 0 {
+		return 0, errors.New("read not supported")
+	}
+
+	n, err := f.c.ReadAt(b, f.p)
+	f.p += int64(n)
+
+	return n, err
+}
+
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekCurrent:
+		if offset == 0 {
+			return f.p, nil
+		}
+
+		f.p += offset
+	case io.SeekStart:
+		f.p = offset
+	case io.SeekEnd:
+		f.p = f.c.size - offset
+	}
+
+	return f.p, nil
+}
+
+func (f *file) Write(p []byte) (int, error) {
+	if f.IsClosed() {
+		return 0, fs.ErrClosed
+	}
+
+	if f.flags&os.O_RDWR != 0 && f.flags&os.O_WRONLY != 0 {
+		return 0, errors.New("read not supported")
+	}
+
+	n, err := f.c.WriteAt(p, f.p)
+	f.p += int64(n)
+
+	return n, err
+}
+
+func (f *file) Close() error {
+	f.Closed = true
+	return nil
+}
+
+func (f *file) Open() error {
+	f.Closed = false
+	return nil
+}
+
+type fileInfo struct {
+	name  string
+	size  int64
+	isDir bool
+}
+
+func newFileInfo(base, fullpath string, size int64) *fileInfo {
+	filename, _ := filepath.Rel(base, fullpath)
+
+	return &fileInfo{
+		name: filename,
+		size: size,
+	}
+}
+
+func (fi *fileInfo) Name() string {
+	return fi.name
+}
+
+func (fi *fileInfo) Size() int64 {
+	return fi.size
+}
+
+func (fi *fileInfo) Mode() os.FileMode {
+	return os.FileMode(0)
+}
+
+func (*fileInfo) ModTime() time.Time {
+	return time.Now()
+}
+
+func (fi *fileInfo) IsDir() bool {
+	return fi.isDir
+}
+
+func (*fileInfo) Sys() interface{} {
+	return nil
+}
+
+type storage struct {
+	files map[string]*file
+}
+
+type content struct {
+	bytes []byte
+	size  int64
+}
+
+func (c *content) WriteAt(p []byte, off int64) (int, error) {
+	l := len(p)
+	if int(off)+l > len(c.bytes) {
+		buf := make([]byte, 2*len(c.bytes)+l)
+		copy(buf, c.bytes)
+		c.bytes = buf
+	}
+
+	n := copy(c.bytes[off:], p)
+	if off+int64(n) > c.size {
+		c.size = off + int64(n)
+	}
+
+	return n, nil
+}
+
+func (c *content) ReadAt(b []byte, off int64) (int, error) {
+	if off >= c.size {
+		return 0, io.EOF
+	}
+
+	l := int64(len(b))
+	if off+l > c.size {
+		l = c.size - off
+	}
+
+	n := copy(b, c.bytes[off:l])
```

Comment:
This cannot be right, l can be less than off here, but that is clearly impossible. I think you have to change `c.bytes[off:l]` to `c.bytes[off:off+l]`. I think there has been mistake testing this code, what was its test coverage?